doc/dev/debugging.rst

```diff
@@ -118,98 +118,52 @@ Will print out something close to the following:
-            "-O3",
-        ]
-    
-        @staticmethod
-        def get_output_filename(infile):
-            path = pathlib.Path(infile)
-            if not path.exists():
-                raise FileNotFoundError("Cannot find {infile}.")
-            return str(path.with_suffix(".o"))
-
-or
-
-.. code-block:: python
-
-    class MyLLCOpt(LLVMIRToObjectFile):
-        """LLVMIR To Object File."""
-    
-        _default_flags = [
-            "--filetype=obj",
-            "--relocation-model=pic",
-            "-O3",
+    DEFAULT_PIPELINES = [
+        (
+            "HLOLoweringPass",
+            [
+                "canonicalize",
+                "func.func(chlo-legalize-to-hlo)",
+                "stablehlo-legalize-to-hlo",
+                "func.func(mhlo-legalize-control-flow)",
+                ...
+            ],
+        ),
+        (
+            "QuantumCompilationPass",
+            [
+                "lower-gradients",
+                "adjoint-lowering",
+                "convert-arraylist-to-memref",
+            ],
+        ),
+        ...
         ]
-    
-In order to actually use this ``PassPipeline``, the user must override the default ``PassPipeline``.
-To do so, use the ``pipelines`` keyword parameter in ``@qjit`` decorator.
-The value assigned to ``pipelines`` must be a list of ``PassPipeline`` that will lower MLIR to binary.
-In this particular case, we are substituting the ``LLVMIRToObjectFile`` pass pipeline with ``MyLLCOpt`` in the default pass pipeline.
-The following will work:
 
 
-.. code-block:: python
-
-    custom_pipeline = [MHLOPass, QuantumCompilationPass, BufferizationPass, MLIRToLLVMDialect, LLVMDialectToLLVMIR, MyLLCOpt, CompilerDriver]
-    
-    @qjit(pipelines=custom_pipeline)
-    def foo():
-        """A method to be JIT compiled using a custom pipeline"""
-        ...
+Here, each item represents a pipeline. Each pipeline has a name and a list of MLIR passes
+to perform. Most of the standard passes are described in the
+`MLIR passes documentation <https://mlir.llvm.org/docs/Passes/>`_. Quantum MLIR passes are
+implemented in Catalyst and can be found in the sources.
 
-Users that are interested in ``PassPipeline`` classes are encouraged to look at the ``compiler.py`` file to look at different ``PassPipeline`` child classes.
+All pipelines are executed in sequence, the output MLIR of each pipeline is stored in
+memory and become available via the ``get_output_of`` method of the ``QJIT`` object.
 
-Printing the IR generated by Pass Pipelines
-==========================================
+Printing the IR generated by Pipelines
+======================================
 
-We won't get into too much detail here, but sometimes it is useful to look at the output of a specific ``PassPipeline``.
+We won't get into too much detail here, but sometimes it is useful to look at the output of a
+specific pass pipeline.
 To do so, simply use the ``get_output_of`` method available in ``QJIT``.
```

Comment:
I'll need to double check, but I believe we do write them to disk. The alternative might be worse: e.g., carry around all string representation of the program in QJIT. I think writing to disk is preferrable. The advantage we gain here is that we don't have to read from this file to parse the IR again, and we could optimize writing to disk by making it asynchronously and non-blocking.

Reply:
Interesting! My thinking is that we should ensure that the "default" path (regular compilation + execution of the user program) is not slowed down, but something like parsing from file is totally fine when the user requests it explicitly since it doesn't happen during "regular" execution.
If the writing to disk doesn't noticeably impact compile time we could keep it "always on", but otherwise I was thinking that asking for intermediate outputs could require the `keep_intermediate` flag to be set.

Reply:
> Interesting! My thinking is that we should ensure that the "regular" path (just compilation + execution of the user program) is not slowed down, but something like parsing from file is totally fine when the user requests it explicitly since it doesn't happen during regular execution.

It used to be that we dump the AST and parse the AST between each of the compilation passes. This re-parsing is stopped now. For large programs, parsing the AST and dumping it could be an issue. This process had to be serialized without the compiler driver. Now we can just dump to a file using thread x and continue compiling in the main thread.

> I was thinking that asking for intermediate outputs could require the keep_intermediate flag to be set.

Yes, this is something that I believe was discussed in the past. But we set to keep the current behaviour.

Reply:
But the current behaviour is still synchronous?

Reply:
Ok then I think either of the two suggestions should be added first. Note that printing intermediate stages (beyond the already available jaxpr/mlir/qir) is something almost no user would use.

Reply:
I agree 100%. Then why not remove it?

Reply:
Hey,

I doubled check and this might not be feasible in the short term. While in theory printing in different threads is definitely possible, the way it is implemented it doesn't make much sense. The printing is managed (scheduled) by MLIR libraries after the execution of each pass. And this means that printing is just done in a small context. If we create a thread in this context, we still need to join in this context, which means there is not really a big advantage.

E.g., 

```
// Printing happens in this callback.
auto printHandler =
                [&](Pass *pass, CatalystIRPrinterConfig::PrintCallbackFn print) -> LogicalResult {
                auto res = pipelineTailMarkers.find(pass);
                if (res != pipelineTailMarkers.end()) {
                    for (const auto &pn : res->second) {
                        std::string outFile = fs::path(std::to_string(pipelineIdx++) + "_" + pn)
                                                  .replace_extension(".mlir");
                        {
                            llvm::raw_string_ostream s{outputs[pn]};
                            print(s);
                        }
                        // If we create a thread here.
                        if (failed(dumpToFile(options, outFile, outputs[pn]))) {
                            return failure();
                        }
                       // We need to join here.
                    }
                }
                return success();
            };
```